File: sdk/javascript/src/projects.ts
import type { CloudPool } from './client';
import type { Project } from './types';

export class ProjectClient {
  constructor(private readonly cp: CloudPool) {}

  /** List all projects for the current user */
  async list(): Promise<Project[]> {
    return this.cp.request<Project[]>('GET', '/api/projects');
  }

  /** Create a new project */
  async create(name: string, description?: string): Promise<Project> {
    return this.cp.request<Project>('POST', '/api/projects', { name, description });
  }

  /** Get a specific project by ID */
  async get(projectId: string): Promise<Project> {
    return this.cp.request<Project>('GET', `/api/projects/${projectId}`);
  }

  /** Delete a project and all associated resources */
  async delete(projectId: string): Promise<void> {
    await this.cp.request('DELETE', `/api/projects/${projectId}`);
  }

  /** Take a snapshot of the project infrastructure state */
  async snapshot(projectId: string, label?: string): Promise<{ snapshotId: string; createdAt: string }> {
    return this.cp.request('POST', `/api/projects/${projectId}/snapshot`, { label });
  }

  /** Restore a project to a previous snapshot */
  async restore(projectId: string, snapshotId: string): Promise<void> {
    await this.cp.request('POST', `/api/projects/${projectId}/restore/${snapshotId}`);
  }
}

File: sdk/python/cloudpool/projects.py
from __future__ import annotations

from typing import Any


class ProjectClient:
    """Project management APIs."""

    def __init__(self, cp):
        self.cp = cp

    def list(self) -> list[dict[str, Any]]:
        """List all projects."""
        return self.cp.request("GET", "/api/projects")

    def create(
        self,
        name: str,
        description: str | None = None,
    ) -> dict[str, Any]:
        """Create a project."""
        return self.cp.request(
            "POST",
            "/api/projects",
            json={
                "name": name,
                "description": description,
            },
        )

    def get(self, project_id: str) -> dict[str, Any]:
        """Get project by ID."""
        return self.cp.request(
            "GET",
            f"/api/projects/{project_id}",
        )

    def delete(self, project_id: str) -> None:
        """Delete project."""
        self.cp.request(
            "DELETE",
            f"/api/projects/{project_id}",
        )

    def snapshot(
        self,
        project_id: str,
        label: str | None = None,
    ) -> dict[str, Any]:
        """Create snapshot."""
        return self.cp.request(
            "POST",
            f"/api/projects/{project_id}/snapshot",
            json={"label": label},
        )

    def restore(
        self,
        project_id: str,
        snapshot_id: str,
    ) -> None:
        """Restore snapshot."""
        self.cp.request(
            "POST",
            f"/api/projects/{project_id}/restore/{snapshot_id}",
        )